distributed/tests/test_worker_state_machine.py

```diff
@@ -245,60 +249,235 @@ def test_executefailure_to_dict():
+            key="x", worker="na", errant_worker=w2.address, stimulus_id="test"
+        )
+        assert len(s.tasks["x"].who_has) == 1
 
-    await f1
+        if as_deps:
+            y2 = c.submit(inc, x, key="y2", workers=[w1.address])
+        else:
+            s.request_acquire_replicas(w1.address, ["x"], stimulus_id="test")
+
+        while w1.tasks["x"].who_has != {w3.address}:
+            await asyncio.sleep(0.01)
+
+    await wait_for_state("x", "memory", w1)
+
+    assert_story(
+        w1.story("request-dep"),
+        [("request-dep", w3.address, {"x"})],
+        # This tests that there has been no attempt to contact w2.
+        # If the assumption being tested breaks, this will fail 50% of the times.
+        strict=True,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_fetch_to_missing(c, s, a, b):
+    """
+    1. task x is a dependency of y
+    2. scheduler calls handle_compute("y", who_has={"x": [b]}) on a
+    3. x transitions released -> fetch -> flight; a connects to b
+    4. b responds it's busy. x transitions flight -> fetch
+    5. The busy state triggers an RPC call to Scheduler.who_has
+    6. the scheduler responds {"x": []}, because w1 in the meantime has lost the key.
+    7. x is transitioned fetch -> missing
+    """
+    x = (await c.scatter({"x": 1}, workers=[b.address]))["x"]
+    b.total_in_connections = 0
+    with freeze_inbound_comms(a):
+        y = c.submit(inc, x, key="y", workers=[a.address])
+        await wait_for_state("x", "fetch", a)
+        # Do not use handle_missing_data, since it would cause the scheduler to call
+        # handle_free_keys(["y"]) on a
+        s.remove_replica(ts=s.tasks["x"], ws=s.workers[b.address])
+        # We used a scheduler internal call, thus corrupting its state.
+        # Don't crash at the end of the test.
+        s.validate = False
+
+    await wait_for_state("x", "missing", a)
+    assert_story(
+        a.story("x"),
+        [
+            ("x", "ensure-task-exists", "released"),
+            ("x", "update-who-has", [], [b.address]),
+            ("x", "released", "fetch", "fetch", {}),
+            ("gather-dependencies", b.address, {"x"}),
+            ("x", "fetch", "flight", "flight", {}),
+            ("request-dep", b.address, {"x"}),
+            ("busy-gather", b.address, {"x"}),
+            ("x", "flight", "fetch", "fetch", {}),
+            ("x", "update-who-has", [b.address], []),  # Called Scheduler.who_has
+            ("x", "fetch", "missing", "missing", {}),
+        ],
+        # There may be a round of find_missing() after this
+        strict=False,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_self_denounce_missing_data(c, s, a):
+    x = c.submit(inc, 1, key="x")
+    await x
+
+    # Manually wipe x from the worker
+    # (Using an endpoint like remove-replicas would inform the scheduler).
+    del a.data["x"]
+    del a.tasks["x"]
+    a.validate_state()
```

Comment:
How is this ever possible? Fox `x` to be removed from tasks and data, this is only possible if it the task is actually forgotten. I don't think we should expect the worker to recover from arbitrary internal state mutations. If this is all about suppressing/slowing down a message to the scheduler, let's think about how we can do this systematically but let's not just mutate state

I don't think we should modify internal state and expect the worker to just deal with this magically.

Reply:
I cannot think from the top of my head of a use case where the scheduler thinks that the worker has a key but it actually doesn't.
I don't think however it's healthy to say "I can't think of anything that could go wrong in this sync over the network, so nothing will go wrong". This failsafe is something that I found in main; I realised it was broken, so I fixed it.

I can figure out another way to make the scheduler and worker fall out of sync - which will almost invariably require monkey-patching of some internals, which IMHO is much worse than this.

The alternative to all this is to trust that this out-of-sync use case is impossible, replace the message with an assertion, and remove the test altogether.

Reply:
As I said, I consider mutating the internals a hard no. I wouldn't consider it unreasonable for there to be another dictionary, set, etc. that tracks something about `x`. We should not encode our perception of the internals in tests. The internal state is much too complex for this.

While this is by no means a clean approach but to simulate network delays during very specific events, we can introduce the network layer, e.g. using https://github.com/dask/distributed/blob/5ca7a5ae32e8a46ad3275d6f000c07498113b785/distributed/utils_test.py#L1900-L1925

Reply:
> I wouldn't consider it unreasonable for there to be another dictionary, set, etc. that tracks something about x

That's the whole reason for the line ``a.validate_state()``.
Anyway, I'll try using the LockedCommPool.

Reply:
I tried using _LockedCommPool but I could not find a way to make it work (the worker would just send the message immediately).
I found an alternative which I think is reasonable: https://github.com/dask/distributed/pull/6342/commits/e6fc28e795005f86e7897f07f23f9194c7935b1d